polkadot/runtime/parachains/src/paras_inherent/mod.rs

```diff
@@ -1189,61 +1241,194 @@ fn filter_backed_statements_from_disabled_validators<T: shared::Config + schedul
 			bc.validity_votes_mut().remove(idx);
 		}
 
-		// If at least one statement was dropped we need to return `true`
-		if indices_to_drop.count_ones() > 0 {
-			filtered = true;
-		}
-
 		// By filtering votes we might render the candidate invalid and cause a failure in
 		// [`process_candidates`]. To avoid this we have to perform a sanity check here. If there
 		// are not enough backing votes after filtering we will remove the whole candidate.
-		if bc.validity_votes().len() < effective_minimum_backing_votes(
-			validator_group.len(),
-			minimum_backing_votes
-		) {
+		if bc.validity_votes().len() <
+			effective_minimum_backing_votes(validator_group.len(), minimum_backing_votes)
+		{
+			log::debug!(
+				target: LOG_TARGET,
+				"Dropping candidate {:?} of paraid {:?} because it was left with too few backing votes after votes from disabled validators were filtered.",
+				bc.candidate().hash(),
+				para_id
+			);
+
 			return false
 		}
 
 		true
 	});
+}
 
-	// Also return `true` if a whole candidate was dropped from the set
-	filtered || backed_len_before != backed_candidates_with_core.len()
+// Check that candidates pertaining to the same para form a chain. Drop the ones that
+// don't, along with the rest of candidates which follow them in the input vector.
+fn filter_unchained_candidates<T: inclusion::Config + paras::Config + inclusion::Config>(
+	candidates: &mut BTreeMap<ParaId, Vec<BackedCandidate<T::Hash>>>,
+	allowed_relay_parents: &AllowedRelayParentsTracker<T::Hash, BlockNumberFor<T>>,
+) {
+	let mut para_latest_head_data: BTreeMap<ParaId, HeadData> = BTreeMap::new();
+	for para_id in candidates.keys() {
+		let maybe_latest_head_data = match <inclusion::PendingAvailability<T>>::get(&para_id)
+			.map(|pending_candidates| {
+				pending_candidates.back().map(|x| x.candidate_commitments().head_data.clone())
+			})
+			.flatten()
+		{
+			Some(head_data) => Some(head_data),
+			None => <paras::Pallet<T>>::para_head(&para_id),
+		};
+		// this cannot be None
+		let latest_head_data = match maybe_latest_head_data {
+			None => {
+				log::warn!(target: LOG_TARGET, "Latest included head data for paraid {:?} is None", para_id);
+				continue
+			},
+			Some(latest_head_data) => latest_head_data,
+		};
+		para_latest_head_data.insert(*para_id, latest_head_data);
+	}
+
+	filter_candidates::<T, _>(candidates, |para_id, candidate| {
+		let Some(latest_head_data) = para_latest_head_data.get(&para_id) else { return false };
```

Comment:
I think in this particular case it helps with code deduplication and I actually think it improves readability. It's useful because all the filtering done here has the particular meaning that we stop once we find a candidate which doesn't satisfy the predicate and drop all the next candidates of the paraid.
I'd much rather have this deduplicated.

The predicates are documented at the function level or when being called, with comments. I can't extract them to separate functions as they capture outer state

Reply:
When reading the code I would expect to easily grasp what is being done at high level  and having the predicates separate helps with that.